Follow up to 295442, add a test and make the ephemeralness bit actually pass through.

megangardner opened the pull request:
#### 432c051f13d5e6e652b7291cd800b5baf0fa44e8
<pre>
Follow up to 295442, add a test and make the ephemeralness bit actually pass through.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295616">https://bugs.webkit.org/show_bug.cgi?id=295616</a>
<a href="https://rdar.apple.com/155390289">rdar://155390289</a>

Reviewed by Wenson Hsieh.

Adding a test.
Also, we didn&apos;t actually determine if a
session was ephemeral and pass that information to
the pasteboard, so adding the code to actually do that.

* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::PlatformPasteboard::write):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::determineDataLifetime):
(WebKit::WebPasteboardProxy::setPasteboardTypes):
(WebKit::WebPasteboardProxy::writeCustomData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
(setClipboardConfigurationPreferences):
(createWebViewForClipboardTests):
(createEphemeralWebViewForClipboardTests):
(TEST(ClipboardTests, EphemeralSessionClipboardHasExpiration)):

Canonical link: <a href="https://commits.webkit.org/297156@main">https://commits.webkit.org/297156@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b0f321c6ecedb0088c29607edc63b04765f2be9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116728 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60968 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112664 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31039 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84177 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113649 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24796 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99686 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64618 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24161 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17824 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60522 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94186 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119518 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37741 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28044 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93140 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92964 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23694 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37995 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15741 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33719 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37637 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43109 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37299 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40638 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39007 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -79,12 +77,38 @@ - (void)writeString:(NSString *)string toClipboardWithType:(NSString *)type
+    TestWebKitAPI::Util::instantiateUIApplicationIfNeeded();
+#endif
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
Nit - `RetainPtr` for smart pointers nowadays.
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -79,12 +77,38 @@ - (void)writeString:(NSString *)string toClipboardWithType:(NSString *)type
+#if PLATFORM(IOS_FAMILY)
+    TestWebKitAPI::Util::instantiateUIApplicationIfNeeded();
+#endif
+    RetainPtr<WKWebsiteDataStore> ephemeralStore = [WKWebsiteDataStore nonPersistentDataStore];
```
```suggestion
    RetainPtr ephemeralStore = [WKWebsiteDataStore nonPersistentDataStore];
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -79,12 +77,38 @@ - (void)writeString:(NSString *)string toClipboardWithType:(NSString *)type
+    TestWebKitAPI::Util::instantiateUIApplicationIfNeeded();
+#endif
+    RetainPtr<WKWebsiteDataStore> ephemeralStore = [WKWebsiteDataStore nonPersistentDataStore];
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -79,12 +77,38 @@ - (void)writeString:(NSString *)string toClipboardWithType:(NSString *)type
+
+    setClipboardConfigurationPreferences(configuration.get());
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 400) configuration:configuration.get()]);
```
```suggestion
    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 400) configuration:configuration.get()]);
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -184,4 +208,25 @@ static void writeMultipleObjectsToPlatformPasteboard()
 
+TEST(ClipboardTests, EphemeralSessionClipboardHasExpiration)
+{
+    auto webView = createEphemeralWebViewForClipboardTests();
```
```suggestion
    RetainPtr webView = createEphemeralWebViewForClipboardTests();
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -184,4 +208,25 @@ static void writeMultipleObjectsToPlatformPasteboard()
+    };
+
+    pasteboardHaveExpirationDate = false;
+
```
Nit - this can be removed

```suggestion
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
```diff
@@ -25,14 +25,17 @@
 #import <WebKit/WKPreferencesPrivate.h>
+#import <WebKit/WKWebsiteDataStoreRef.h>
 #import <WebKit/_WKFeature.h>
+#import <pal/spi/mac/NSPasteboardSPI.h>
```
This will need to be behind a `PLATFORM(MAC)`, or you could add guards around the SPI header itself.

```suggestion

#if PLATFORM(MAC)
#import <pal/spi/mac/NSPasteboardSPI.h>
#endif
```

whsieh reviewed (approved):
r=mews once iOS-family builds are green

megangardner committed:
Follow up to 295442, add a test and make the ephemeralness bit actually pass through.
https://bugs.webkit.org/show_bug.cgi?id=295616
rdar://155390289

Reviewed by Wenson Hsieh.

Adding a test.
Also, we didn't actually determine if a
session was ephemeral and pass that information to
the pasteboard, so adding the code to actually do that.

* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::PlatformPasteboard::write):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::determineDataLifetime):
(WebKit::WebPasteboardProxy::setPasteboardTypes):
(WebKit::WebPasteboardProxy::writeCustomData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ClipboardTests.mm:
(setClipboardConfigurationPreferences):
(createWebViewForClipboardTests):
(createEphemeralWebViewForClipboardTests):
(TEST(ClipboardTests, EphemeralSessionClipboardHasExpiration)):

Canonical link: https://commits.webkit.org/297156@main
